### DISCUSSION TITLE
Fixes for 7.0.0

### DISCUSSION THREAD

**Discussion by philipphofmann** [getsentry/sentry-cocoa#877] [General]
This is an informal place to collect things to improve or fix for the major version, that are breaking changes. Once we start with the next major version we are going to create issues for them.

**Comment by philipphofmann** [comment]
Remove `isEqual`, `isEqualToScope` and `hash` from `SentryScope`. 

The `SentryScope` is mutable and using these methods without care can cause issues especially with hash tables. These methods were added with https://github.com/getsentry/sentry-cocoa/pull/743/ and from a user's perspective, it doesn't really make sense to have them. Removing them simplifies the API and prevents the above-mentioned issues.

**Comment by philipphofmann** [comment]
Return `nil` if the `jsonObject` passed to [`Session.initWithJson`](https://github.com/getsentry/sentry-cocoa/blob/d98956472bd2f90848f451de164ee58049148cd4/Sources/Sentry/SentrySession.m#L27) is not valid.

**Reply by philipphofmann** [reply]
Fixed with https://github.com/getsentry/sentry-cocoa/pull/939

**Comment by philipphofmann** [comment]
Replace [`SentryHub.getScope`](https://github.com/getsentry/sentry-cocoa/blob/1610082263e9d27533de2aa7b80d89b154a9176d/Sources/Sentry/Public/SentryHub.h#L131) with a readonly property.

**Reply by philipphofmann** [reply]
See https://github.com/getsentry/sentry-cocoa/pull/975

**Comment by philipphofmann** [comment]
Set default of [maxCacheItems](https://github.com/getsentry/sentry-cocoa/issues/899) to 30, like in Java and .NET

**Reply by philipphofmann** [reply]
See https://github.com/getsentry/sentry-cocoa/pull/1017

**Comment by philipphofmann** [comment]
Protocols for Client, Hub, and Transport.

**Comment by philipphofmann** [comment]
Remove [`SentrySDK.currentHub`](https://github.com/getsentry/sentry-cocoa/blob/ff37e5dbdcc47e53144d00224a236c0c6b3dafb2/Sources/Sentry/Public/SentrySDK.h#L22) and [`SentrySDK.setCurrentHub`](https://github.com/getsentry/sentry-cocoa/blob/ff37e5dbdcc47e53144d00224a236c0c6b3dafb2/Sources/Sentry/Public/SentrySDK.h#L32).

RN needs `SentrySDK.currentHub`, see https://github.com/getsentry/sentry-react-native/blob/8fc04313c4143ddd3ef6302d7ec08042924bf66d/ios/RNSentry.m#L149-L157, but we could replace this with a method on `SentrySDK`.

**Reply by philipphofmann** [reply]
Also add `startSession` and `endSession` when removing the `currentHub`, see https://github.com/getsentry/develop/issues/248

**Comment by philipphofmann** [comment]
Add a thread-safe implementation of a dictionary to expose the dict instead of having methods as `setTag` and `removeTag`, for example, on the Scope or SpanContext.

**Reply by bruno-garcia** [reply]
We don't want to expose a Map to users though right? In most SDks Sentry doesn't even let user read the scope. The more API surface we add here the harder to evolve it becomes (like synchronizing the scope on change for example).

**Reply by bruno-garcia** [reply]
Not against it but might be that we just keep that internal. On Java for example it seems we return the actual reference to the dictionary so users can do spanContext.getTags().whateverMapClassExposes.. hold that reference as a field or whatever. It's just very leaky. Ideally it would be a readonly interface or just returns a key/value iterator

**Reply by philipphofmann** [reply]
Yes, I agree making it readonly makes sense.

**Comment by philipphofmann** [comment]
Replace [SentryExcption.thread](https://github.com/getsentry/sentry-cocoa/blob/10d7a5b2816201dde26a4fe630fca71dd9f6f06f/Sources/Sentry/Public/SentryException.h#L42) with `threadId` and `stacktrace`, as described in the [exception interface](https://develop.sentry.dev/sdk/event-payloads/exception/). Internally we already serialize according to the interface, https://github.com/getsentry/sentry-cocoa/blob/10d7a5b2816201dde26a4fe630fca71dd9f6f06f/Sources/Sentry/SentryException.m#L28-L29

**Reply by philipphofmann** [reply]
See https://github.com/getsentry/sentry-cocoa/pull/1026

**Comment by philipphofmann** [comment]
Use a class for [mechanism metadata](https://develop.sentry.dev/sdk/event-payloads/exception/#meta-information) instead of a Dictionary on `SentryMechanism`. Also, move `SentryNSError` to this class.

**Reply by philipphofmann** [reply]
See https://github.com/getsentry/sentry-cocoa/pull/1038

**Comment by philipphofmann** [comment]
7.0.0 is out now. Put the fixes for the next major in 8.0.0.

**Reply by armcknight** [reply]
And also 8.0.0 is out now, so [that discussion](https://github.com/getsentry/sentry-cocoa/discussions/1094) is also now closed. See the [new one for 9.0.0](https://github.com/getsentry/sentry-cocoa/discussions/2946)!
